File: opensrp-chw-referral/src/main/java/org/smartregister/chw/referral/presenter/BaseIssueReferralPresenter.java
package org.smartregister.chw.referral.presenter;

import android.util.Log;

import org.apache.commons.lang3.tuple.Triple;
import org.jetbrains.annotations.Nullable;
import org.smartregister.chw.referral.contract.BaseIssueReferralContract;
import org.smartregister.chw.referral.domain.MemberObject;
import org.smartregister.chw.referral.model.AbstractIssueReferralModel;
import org.smartregister.chw.referral.util.Constants;
import org.smartregister.chw.referral.util.DBConstants;

import java.lang.ref.WeakReference;
import java.util.List;

import timber.log.Timber;

public class BaseIssueReferralPresenter implements BaseIssueReferralContract.Presenter, BaseIssueReferralContract.InteractorCallBack {
    protected BaseIssueReferralContract.Interactor interactor;
    private MemberObject memberObject;
    private String baseEntityID;
    private WeakReference<BaseIssueReferralContract.View> viewReference;
    private Class<? extends AbstractIssueReferralModel> viewModelClass;

    public BaseIssueReferralPresenter(String baseEntityID, BaseIssueReferralContract.View view, Class<? extends AbstractIssueReferralModel> viewModelClass, BaseIssueReferralContract.Interactor interactor) {
        viewReference = new WeakReference<>(view);
        this.interactor = interactor;
        this.baseEntityID = baseEntityID;
        this.viewModelClass = viewModelClass;
    }

    @SuppressWarnings("unchecked")
    @Override
    public Class<? extends AbstractIssueReferralModel> getViewModel() {
        return viewModelClass;
    }

    @Override
    public String getMainCondition() {
        try {
            return Constants.TABLES.FAMILY_MEMBER + "." + DBConstants.KEY.BASE_ENTITY_ID + " = '" + memberObject.getBaseEntityId() + "'";
        } catch (Exception e) {
            Timber.e(e);
        }
        return null;

    }

    @Override
    public String getMainTable() {
        return Constants.TABLES.FAMILY_MEMBER;
    }

    @Override
    public void fillClientData(@Nullable MemberObject memberObject) {
        if (memberObject != null && getView() != null) {
            getView().setProfileViewWithData();
        }
    }

    @Override
    public void initializeMemberObject(MemberObject memberObject) {
        this.memberObject = memberObject;
    }


    @Override
    public void saveForm(String jsonString) {
        try {
            interactor.saveRegistration(jsonString, this);
        } catch (Exception e) {
            Timber.e(Log.getStackTraceString(e));
        }
    }


    @Override
    public void onUniqueIdFetched(Triple<String, String, String> triple, String entityId) {
//        implement
    }

    @Override
    public void onNoUniqueId() {
//        implement
    }

    @Override
    public void onRegistrationSaved() {
//        implement
    }

    @Override
    public void registerViewConfigurations(List<String> list) {
//        implement
    }

    @Override
    public void unregisterViewConfiguration(List<String> list) {
//        implement
    }

    @Override
    public void onDestroy(boolean b) {
//        implement
    }

    @Override
    public void updateInitials() {
//        implement
    }


    private BaseIssueReferralContract.View getView() {
        if (viewReference != null)
            return viewReference.get();
        else
            return null;
    }

    public String getBaseEntityID() {
        return baseEntityID;
    }
}

File: opensrp-chw-referral/src/main/java/org/smartregister/chw/referral/presenter/BaseIssueReferralPresenter.kt
package org.smartregister.chw.referral.presenter

import android.util.Log
import androidx.lifecycle.ViewModel
import com.nerdstone.neatformcore.domain.model.NFormViewData
import org.apache.commons.lang3.tuple.Triple
import org.json.JSONObject
import org.smartregister.chw.referral.contract.BaseIssueReferralContract
import org.smartregister.chw.referral.domain.MemberObject
import org.smartregister.chw.referral.model.AbstractIssueReferralModel
import org.smartregister.chw.referral.util.Constants
import org.smartregister.chw.referral.util.DBConstants
import timber.log.Timber
import java.lang.ref.WeakReference
import java.util.*

open class BaseIssueReferralPresenter(
    val baseEntityID: String,
    view: BaseIssueReferralContract.View,
    private val viewModelClass: Class<out AbstractIssueReferralModel>,
    protected var interactor: BaseIssueReferralContract.Interactor
) : BaseIssueReferralContract.Presenter, BaseIssueReferralContract.InteractorCallBack {

    var memberObject: MemberObject? = null
    private var viewReference = WeakReference(view)

    override fun getView(): BaseIssueReferralContract.View? {
        return viewReference.get()
    }

    override fun <T> getViewModel(): Class<T> where T : ViewModel, T : BaseIssueReferralContract.Model {
        return viewModelClass as Class<T>
    }

    override fun getMainCondition() =
        "${Constants.Tables.FAMILY_MEMBER}.${DBConstants.Key.BASE_ENTITY_ID}  = '$baseEntityID'"

    override fun getMainTable() = Constants.Tables.FAMILY_MEMBER

    override fun fillClientData(memberObject: MemberObject?) {
        if (getView() != null) {
            getView()?.setProfileViewWithData()
        }
    }

    override fun initializeMemberObject(memberObject: MemberObject?) {
        this.memberObject = memberObject
    }

    override fun saveForm(valuesHashMap: HashMap<String, NFormViewData>, jsonObject: JSONObject) {
        try {
            interactor.saveRegistration(baseEntityID, valuesHashMap, jsonObject, this)
        } catch (e: Exception) {
            Timber.e(Log.getStackTraceString(e))
        }
    }

    override fun onUniqueIdFetched(triple: Triple<String, String, String>, entityId: String) = Unit

    override fun onNoUniqueId() = Unit

    override fun onRegistrationSaved() = Unit

}